### DISCUSSION TITLE
Switched to different i2s mic, now only getting clipped / no audio. Config?

### DISCUSSION THREAD

**Discussion by musiq021** [pschatzmann/arduino-audio-tools#1557] [Q&A] [answered]
Had to change mics in my project due to low supply.

Switched from the ICS-43432 to the SPH0645LM4H-B

On at least one of the boards, I'm not getting any audio, possible reflow issue, but one of them IS "working"...

The mic produces no audio unless I shout into it, at which point it only picks up clipped audio in square waves / clipping ( see screenshot )

I want to troubleshoot if maybe it's because I configured it to transmit on the wrong channel, but I don't understand how to select the i2s input channel in audio-tools.

My other suspicion is that maybe it's misbehaving because I didn't follow the decoupling in the app notes exactly. It calls for a 0.1uf and 200pf whereas I only have the 0.1uf.
![Screenshot 2024-05-08 at 09 03 33](https://github.com/pschatzmann/arduino-audio-tools/assets/1408450/b09c8877-94f0-4afc-b428-20e1286352f0)

![Screenshot 2024-05-08 at 07 27 30](https://github.com/pschatzmann/arduino-audio-tools/assets/1408450/017082cc-b6eb-403a-a7a3-b989ad174883)
![Screenshot 2024-05-08 at 07 27 07](https://github.com/pschatzmann/arduino-audio-tools/assets/1408450/35805dc6-a189-4c9b-ad36-ec452d9bcbe5)
`    auto cfg = i2s.defaultConfig(RXTX_MODE);
    cfg.pin_ws = i_ws;
    cfg.pin_bck = i_bck;
    cfg.pin_data = i_data;
    cfg.pin_data_rx = i_rx;
    cfg.auto_clear = true; //supposed to help with pops. doesn't seem to
    i2s.begin(cfg);`

**Comment by musiq021** [comment]
I managed to bodge the select pin to GND instead and it resulted in no output whatsoever, so I think I can rule out the channel selection being the issue ( which would make sense since I shouldn't expect ANY result if the wrong channel were selected )

**Reply by pschatzmann** [reply]
I2S is a stereo protocol, so I don't think you would need to bother about the channel selection. 
I suggest that you try different i2s_format settings

**Comment by musiq021** [comment]
UGH, just found this, looks like there is a timing issue with this mic on the ESP32, https://hackaday.io/project/162059-street-sense/log/160705-new-i2s-microphone.

Someone in the comments suggested a solution, looking it over to see if there's a way to implement it with audio-tools...

**Reply by pschatzmann** [reply, marked as answer]
Seems that you just need to set the channel_format to left or right only and the bits_per_sample to 32 and the i2s_format to I2S_MSB_FORMAT

**Reply by musiq021** [reply]
Wow. Thanks Phil. As always.
Only issue I have now is that since it is recording in 32bps, all previous recordings playback at double speed because the wav decoder doesn't seem to be picking up on the format and passing it to the i2s object despite calling     decoder.setNotifyAudioChange(i2s);

Is there more to it than that?

**Reply by pschatzmann** [reply]
Set the logging to info to check what is happening.
Did you try with the latest corrections of the library ? I think during some time the i2s update was not working properly...

**Reply by musiq021** [reply]
I updated the library, and now by default it plays the 16bps recordings normally, but plays the 32bps recordings as pure static.

I: Playing /waves/cur/240508-090634.wav
[I] StreamCopy.h : 70 - buffer_size=1024
[I] AudioEncoded.h : 321 - virtual void audio_tools::EncodedAudioOutput::addNotifyAudioChange(audio_tools::AudioInfoSupport&)
[I] StreamCopy.h : 147 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops ...

**Reply by pschatzmann** [reply]
Can you provide your sketch that is not working as expected ?

**Comment by musiq021** [comment]
its part of a larger project, but here is the relevant class :

```

 #pragma once
// #define URLSTREAMS
#include "AudioTools.h" // https://github.com/pschatzmann/arduino-audio-tools  : https://github.com/pschatzmann/arduino-audio-tools/discussions 
#include "AudioLibs/AudioSourceSD.h" //sdfat?
#include "AudioCodecs/CodecWAV.h"
#include <AudioFilter/Equilizer.h>
#include <etFiles.h>
#include <Tasker.h>
#include <io.h>
#include <prefs.h>
#include "soc/i2s_reg.h"

class soundSystem {
public:
  enum AudioStates { PLAYING_SOUND, RECORDING, STOPPED, STREAMING, PLAYING_STATION };
  float defVolPlayFactor = 1;
  uint8_t defMicGain = 1; //V02 = 11
  float volMaxGain = 1.5;
  float volMaster = 1;
  float volPlayFactor = 1;
  float volStationFactor = 0.7;

  
  soundSystem() {
    AudioLogger::instance().begin(Serial, AudioLogger::Info); // Set to Info to debug audio issues. Set to warning otherwise.

  }
void init(){


    auto cfg = i2s.defaultConfig(RXTX_MODE);
    cfg.pin_ws = i_ws;
    cfg.pin_bck = i_bck;
    cfg.pin_data = i_data;
    cfg.pin_data_rx = i_rx;
    cfg.i2s_format = I2S_MSB_FORMAT; // added to support  SPH0645
    cfg.fixed_mclk = 0;
    // cfg.channel_format = I2S_CHANNEL_FMT_ONLY_RIGHT;
    // cfg.auto_clear = true; //supposed to help with pops. doesn't seem to
    i2s.begin(cfg);

    auto recVolCfg = recVol.defaultConfig(); // need to amplify microphone
    recVolCfg.allow_boost = 1;
    recVolCfg.volume = defMicGain; //20
    recVol.begin(recVolCfg);

    decoder.addNotifyAudioChange(i2s);
    decoder.begin(cfg);

    eqConfig.freq_low = 30;
    eqConfig.freq_high = 3500;
    eqConfig.gain_low = 0.2;
    eqConfig.gain_medium = .33;
    eqConfig.gain_high = 1;
    // eqConfig.bits_per_sample =16;
    // eQ.setAudioInfo(recInfo);
    eQ.begin(eqConfig);

    auto playVolCfg = mainVol.defaultConfig();
    playVolCfg.volume = volMaster;
    playVolCfg.allow_boost = 1;
    mainVol.begin(playVolCfg);

    wavEnc.begin(recInfo);
    copier.begin();
    fileRec.close();

}
  void closeFiles(){
    filePlay.close();
    fileRec.close();
  }

  void playStop() {
    if (filePlay) {
      if (audioState == PLAYING_STATION) {
        activeStationPos = filePlay.position();
      }
      filePlay.flush();
      filePlay.close();
    }
      //i2s.writeSilence(1024);
      audioState = STOPPED;
  }

  void playStart(String soundPath, bool skipFileInit = false) {
    static String soundPathLast = "";
    // decoder.addNotifyAudioChange(i2s);
    // streamStop();
    Log.infoln("Playing %s", soundPath.c_str());
    // i2s.setAudioInfo(recInfo);
    if(soundPathLast == soundPath && filePlay){//if the same file is being played, just seek to the beginning
      skipFileInit = true;
      filePlay.seek(0);
    }
    if(!skipFileInit){
      if (filePlay) { 
        playStop(); 
      } 
      filePlay = SD.open(soundPath);
    }

    soundPathLast = soundPath;
    // decoder.addNotifyAudioChange(i2s);
    decoder.begin();
    copier.begin(decoder, filePlay);
  }


  bool recStart(File newFileRec) {
    streamStop();
    playStop();
    i2s.setAudioInfo(recInfo);
    fileRec = newFileRec;
    if (fileRec) {
      sprintf(curRecName, fileRec.name());
      Serial.printf("%s Opened to record. \n", curRecName);
      wavEnc.begin(recInfo);
      copier.begin(recVol, i2s);
      audioState = RECORDING;
      return true;
    }
    else {
      Serial.printf("Failed to write to %s", curRecName);
      return false;
    }
  }

  void recStop() {
    if (fileRec) {
      fileRec.flush();
      Serial.print("File has ");
      Serial.print(fileRec.size());
      Serial.println(" bytes");
      fileRec.close();
      audioState = STOPPED;
    }
  }

  char* latestRecPath() {
    return curRecName;
  }

AudioStates getState(){
  return audioState;
}

  StreamCopy copier{ decoder, filePlay };

private:
  File filePlay;
  File fileRec;
  char curRecName[48];

  AudioStates audioState = STOPPED;

  AudioInfo recInfo{ 22050, 1, 32 }; // 22.05kHz, 1 channel, 16 bits per sample
  
  I2SStream i2s;
  VolumeStream mainVol{i2s};
  Equilizer3Bands eQ{mainVol};
  EncodedAudioStream decoder{ &eQ, new WAVDecoder() };
  EncodedAudioStream wavEnc{ &fileRec, new WAVEncoder() };
  VolumeStream recVol{ wavEnc };

 ConfigEquilizer3Bands eqConfig;

};

```

**Reply by musiq021** [reply]
Hey Phil,
I updated the library and haven't been able to get the mic working again since.
    config.bits_per_sample = 32;
    config.i2s_format = I2S_MSB_FORMAT; 

    config.channel_format = I2S_CHANNEL_FMT_ONLY_RIGHT; ( Outputs spurious values ie.: 

2006515712
-1931509760
-1804468224
-1925906432
-2114158592
2085715968
-2076082176
-1919909888

  config.channel_format = I2S_CHANNEL_FMT_ONLY_LEFT outputs 1's unless I yell, which oscillates between 1 and 0.

So it seems like some of the right channel data is bleeding into the left channel.

Every forum post I see on the topic of ESP32 issues with the shp0645 either goes down some bit shifting rabbit hole, or suggests this as a fix : 
   i2s_driver_install(i2s_num, &i2s_config, 0, NULL);   //install and start i2s driver
   REG_SET_BIT(  I2S_TIMING_REG(i2s_num),BIT(9));   /*  #include "soc/i2s_reg.h"   I2S_NUM -> 0 or 1*/
   REG_SET_BIT( I2S_CONF_REG(i2s_num), I2S_RX_MSB_SHIFT);

So I've tried calling the two REG_SET_BIT calls after i2s.begin() which compiles but has no effect.

Any thoughts?

**Reply by pschatzmann** [reply]
If this was working at some point of time, I don't see why it would not work any more.
Try to [isolate the issue ](https://github.com/pschatzmann/arduino-audio-tools/wiki/External-ADC)and check the log if the bits per sample are handled correctly.

**Comment by musiq021** [comment]
I've started from scratch, outputting straight to CSV. There seems to be some sort of offset / timing issue.

( left channel 1 oscillates at high volume, suggesting it is the MSB from channel 2? )
MSB, 2CH 32bps
1 [00000000000000000000000000000001], -230490112 [11111110011000000000000000000000]
1 [00000000000000000000000000000001], -230850560 [11111110010000000000000000000000]
1 [00000000000000000000000000000001], -231145472 [11111110001100000000000000000000]
1 [00000000000000000000000000000001], -231604224 [11111110000110000000000000000000]
1 [00000000000000000000000000000001], -231899136 [11111110000011000000000000000000]
1 [00000000000000000000000000000001], -231997440 [11111110000001000000000000000000]
1 [00000000000000000000000000000001], -231800832 [11111110000010000000000000000000]
( appears the same as Phillips )
MSB, 1CH 32bps
-116981760 [11111001000111000000000000000000]
-116916224 [11111001001000000000000000000000]
-116883456 [11111001001001000000000000000000]
-116867072 [11111001001001100000000000000000]
-116473856 [11111001010000000000000000000000]
-116457472 [11111001010000100000000000000000]

**Reply by musiq021** [reply]
I'd have given up and switched mics by this point if I didn't already have 2 dozen boards using it ( I overestimated manufacturer adherence to standards when i confidently swapped it out ) and if there weren't so many discussions elsewhere of people with the same issue, most of them resolve it with the register calls, but they have no effect.

I'm willing to accept that this may be a SPH0645 + ESP32-S2 issue and leave this as a cautionary tale for anyone else to avoid this mic at all costs.

For reference, from around the web:
https://reversatronics.blogspot.com/2020/06/i2s-microphones-on-esp32-how-high-can-i.html
https://stackoverflow.com/questions/66477892/does-anyone-have-experience-with-a-sph0645i2s-microphone-and-esp32-s2
https://www.esp32.com/viewtopic.php?t=4997
https://www.esp32.com/viewtopic.php?t=35402
https://www.reddit.com/r/esp32/comments/utt7ul/esp32_and_sph0645_i2s_mems_microphone_no_output/
https://www.youtube.com/watch?v=3g7l5bm7fZ8

**Reply by pschatzmann** [reply]
I guess if you want to add these register calls you would need to do this at the right place in the [I2S Driver](https://github.com/pschatzmann/arduino-audio-tools/blob/main/src/AudioI2S/I2SESP32.h). But this will not work in the new I2S API that is coming in the next release

**Reply by musiq021** [reply]
Managed to get it to "work" again by using the NumberFormatConverter from 32 to 16 bps. However, it's quick to clip / lacking in dynamic range, and the output values are centered around -3800 instead of 0. So there seems to be an unwanted offset.
( The register changes didn't seem to have any effect one way or the other )

**Reply by musiq021** [reply]
For anyone else still playing along. This is what the DC offset looks like for the SPH0645 ( this is another issue I've found many reports of ) . I was able to rectify it using Phil's ConverterAutoCenter . 
![SPH0645-DC-Offset](https://github.com/pschatzmann/arduino-audio-tools/assets/1408450/d456a32b-bdf5-4950-a47b-f3b69413d3be)
